### PR TITLE
Fix bad interaction between promotion and incremental builds on OSX

### PR DESCRIPTION
**Pull request by diml**
Previously we were only using mtime but it's not precise enough on OSX.

### CODE REVIEW

**Comment by rgrinberg** [comment]
CI still fails:

```
        diff (internal) (exit 1)
(cd _build/default && /usr/bin/diff -u test/blackbox-tests/test-cases/promote/run.t test/blackbox-tests/test-cases/promote/run.t.corrected)
--- test/blackbox-tests/test-cases/promote/run.t	2018-01-29 16:53:36.000000000 +0000
+++ test/blackbox-tests/test-cases/promote/run.t.corrected	2018-01-29 16:53:58.000000000 +0000
@@ -17,11 +17,8 @@
 
   $ echo titi > x
   $ $JBUILDER build --root . -j1 --diff-command false @blah --auto-promote 2>&1 | sed 's/.*false.*/DIFF/'
-            sh (internal) (exit 1)
-  DIFF
-  Promoting _build/default/x.gen to x.
   $ cat x
-  toto
+  titi
   $ $JBUILDER build --root . -j1 --diff-command false @blah
   $ cat x
-  toto
+  titi
```

With a different message now though.

**Comment by diml** [comment]
And obviously, I can't reproduce it on my macbook...

**Comment by diml** [comment]
OSX is very frustrating:

```
$ echo toto > x; stat -s x |tr " " "\n" > s1; cat x > y; cat y > x; stat -s x |tr " " "\n" > s2; diff -u s1 s2
--- s1	2018-01-30 08:35:55.000000000 -0500
+++ s2	2018-01-30 08:35:55.000000000 -0500
@@ -6,7 +6,7 @@
 st_gid=1068632065
 st_rdev=0
 st_size=5
-st_atime=1517319354
+st_atime=1517319355
 st_mtime=1517319355
 st_ctime=1517319355
 st_birthtime=1517319048
```

**Comment by diml** [comment]
Ok, I made sure that we delete the destination file before copying to force a new inode to be allocated. I also removed some old stuff about updated files, which dates from when we were using timestamps for incremental compilation.

**Comment by diml** [comment]
Ok, so basically we just can't rely on `stat` on OSX. That means that we must re-hash files outside of _build on every run. I need to change the code so that we only do that on OSX. I'm also trying to see if rehashing everything is OK with a faster hash.

**Comment by diml** [comment]
@rgrinberg, i've made the minimal necessary changes to improve the behaviour and get the tests to pass. It's ready for review.

I'll look later into the idea of systematically rehashing external files and using a faster hash implementation.

File: src/utils.ml [hunk 1]
```diff
@@ -196,10 +196,7 @@ module Cached_digest = struct
               };
       digest
 
-  let remove fn =
-    match Hashtbl.find cache fn with
-    | None -> ()
-    | Some file -> file.timestamp_checked <- false
+  let remove fn = Hashtbl.remove cache fn
```
Comment (rgrinberg) [inline src/utils.ml]: Would you mind explaining this change? Why was setting it to false a problem before?

Comment (diml) [reply src/utils.ml]: - it was useless given that we were doing that for the targets of a rule that we always delete before running the rule anyway
- it didn't effectively removed entries from the on-disk cache

**Pull request merged**